src/parser.satyg

```diff
@@ -119,9 +119,9 @@ end = struct
   let run p s =
     let initial-state = (| input = s |) in
     match p initial-state with
-    | (Some (_, a), _, []) ->
+    | (Some (_, a), _, _) ->
       Result.ok a
-    | (_, _, es) ->
+    | (None, _, es) ->
```

Comment:
Aw `Parser.run` seems completely broken 😞 
But I am not sure about this fix.

In principle, after running `Parser.run` we will be in one of the following three states:

1. Got a result, no errors
2. Got a result, but some errors were reported
3. No result, possibly with some errors

The second case occurs for example when the parser automatically recovers from the "missing closing brace '}'" error, as noted in https://github.com/nyuichi/satysfi-base/blob/d951d3fa325548e46f3a9f0b08502cbeaad78c2e/src/parser.satyg#L116.
However, the return type of `Parser.run` only models the first and third cases.
I seem to have overlooked this because the exhaustedness checker reported nothing.

I think the neat fix should change the return type of `Parser.run` to include another variant.
But if we try to avoid such a breaking change, this proposed change should be the best option.

Reply:
あーー sorry I misunderstood your change. I though the original code was like 
```
    match p initial-state with
     | (Some (_, a), _, []) ->
       Result.ok a
     | (None, _, es) ->
       Result.err es
```
Probably the behavior of the original code was intensional. It just classifies the case 1 of the above comment as Ok and case 2 and 3 as Err. Since that the error list is non-empty means something was wrong during parsing, this seems natural to me.

What confuses you should be due to the implementation of `alt`: https://github.com/nyuichi/satysfi-base/blob/master/src/parser.satyg#L179
`alt p1 p2` first runs the parser `p1`. If `p1` fails with error reports `e1` it then runs `p2`. And if `p2` succeeds with a value `v` with error reports `e2`, while we expect the whole result should be `(v, e2)`, it returns `(v, List.append e1 e2)`. This is very counterintuitive and should be fixed, because `e1` is unrelated to parsing of `v`.

Could you update the PR to follow this logic? Or I can make changes by myself.